wallet: Improve log output for errors during load

gwillen opened the pull request:
When loading the wallet, display the entire path in error messages, instead of
the name (which, for the default wallet, is the empty string.)

When an exception occurs during wallet loading, display e.what() if possible,
instead of nothing.

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -4097,7 +4097,7 @@ bool CWallet::Verify(interfaces::Chain& chain, const WalletLocation& location, b
 std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain, const WalletLocation& location, uint64_t wallet_creation_flags)
 {
-    const std::string& walletFile = location.GetName();
+    const std::string& walletFile = (location.GetName() == "" ? "[default wallet]" : location.GetName());
```
```suggestion
    const std::string& log_wallet_name = (location.GetName() == "" ? "[default wallet]" : location.GetName());
```

nit: I'd feel more safe if this was renamed to `log_wallet_name`

gwillen replied:
that makes sense, I will do that.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -4097,7 +4097,7 @@ bool CWallet::Verify(interfaces::Chain& chain, const WalletLocation& location, b
 std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain, const WalletLocation& location, uint64_t wallet_creation_flags)
 {
-    const std::string& walletFile = location.GetName();
+    const std::string& walletFile = (location.GetName() == "" ? "[default wallet]" : location.GetName());
```
nit, this seems more readable: 

`const std::string& walletFile = location.GetName().empty() ? "[default wallet]" : location.GetName();`

alternatively, why not just have `[default wallet]` be the thing set in the constructor for `WalletLocation` and just remove the need for this special casing?

gwillen replied:
I was uncomfortable with modifying it in WalletLocation because I don't fully understand all the uses that the 'wallet name' is put it, and it seems like sometimes it's used as a path/filename.

gwillen replied:
Done.

instagibbs commented on src/wallet/walletdb.cpp:
```diff
@@ -422,8 +422,17 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+        return false;
     } catch (...)
     {
+        if (strErr == "") {
```
nit: `strErr.empty()`

gwillen replied:
Done.

instagibbs commented on src/wallet/walletdb.cpp:
```diff
@@ -422,8 +422,17 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
         }
+    } catch (const std::exception& e)
+    {
+        if (strErr == "") {
```
nit: `strErr.empty()`

gwillen replied:
Done.

Sjors commented:
tACK e5637b1

I now get a flood of this with one my broken wallets:
```sh
2019-02-27T08:32:30Z [default wallet] CDataStream::read(): end of data: unspecified iostream_category error
2019-02-27T08:32:30Z [default wallet] CDataStream::read(): end of data: unspecified iostream_category error
2019-02-27T08:32:30Z [default wallet] CDataStream::read(): end of data: unspecified iostream_category error
```

gwillen commented:
@Sjors Yep, that's the same thing I get -- at least in my case, each of those is a 'keymeta' entry, whose value is missing the final has_key_origin field of CKeyMetaData.

Sjors commented:
It's fine. This is the type of error that should really _never_ happen in the real world, so indeed when it does, the more information the better.

instagibbs commented:
@Sjors frankly these type of changes are mostly for developer's sanity. I've run into similar issues with wallet entries and felt compelled to upstream changes.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/15491/commits/e5637b1dcaf6282e159dce5299a6a365df20454c

promag commented on src/wallet/walletdb.cpp:
```diff
@@ -422,8 +422,17 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
                 strType != "minversion" && strType != "acentry") {
             wss.m_unknown_records++;
         }
+    } catch (const std::exception& e)
```
nit, `{` here.

gwillen replied:
Fixed. It's annoying that the linter doesn't catch stuff like this. Then the wrong version can crop up in the code, and people like me can copy it. :-)

promag reviewed (commented):
Does it make sense to commit some broken wallets to the repo?

gwillen commented:
The idea of committing a broken wallet as a test case makes sense to me, but (1) I seem to recall I've heard people say in past cases that committing a binary wallet file to the repo is not reasonable even as test data, and (2) I definitely don't know how to go about minimizing such a test case, which seems both necessary and painful.

gwillen commented:
Rebased.

promag commented:
Restarted travis job 6 - failed on commit 4e7fea2 with:
```
47/118 - rpc_psbt.py failed, Duration: 4 s
stdout:
2019-03-06T05:02:08.027000Z TestFramework (INFO): Initializing test directory /tmp/test_runner_₿_🏃_20190306_045534/rpc_psbt_69
2019-03-06T05:02:11.006000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/test_framework.py", line 175, in main
    self.run_test()
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/rpc_psbt.py", line 149, in run_test
    assert_raises_rpc_error(-22, "TX decode failed", self.nodes[0].converttopsbt, signedtx['hex'])
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/util.py", line 105, in assert_raises_rpc_error
    assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
AssertionError: No exception raised
```

promag commented:
@gwillen have you seen #15334? Do you think logging the absolute path makes it more clear - instead of "[default wallet]"?

gwillen commented:
Yeah, I think switching to absolute path would make sense. I assume I just need to switch from GetName to GetPath. Will try it out.

gwillen committed:
wallet: Improve log output for errors during load

When loading the wallet, display the entire path in error messages, instead of
the name (which, for the default wallet, is the empty string.)

When an exception occurs during wallet loading, display e.what() if possible,
instead of nothing.

gwillen commented:
Switched to GetPath().string() instead of GetName(). Removed the rename to log_wallet_name, in the interest of minimizing the size of the diff, since it is once again a wallet file name. (I can reinstate the rename if preferred.)

gwillen commented:
Oh, no, this didn't do what I wanted at all. On further inspection of #15334, it doesn't actually log the absolute path, it just logs the filename, but more importantly it doesn't use WalletLocation::GetPath, but it has some other place it's getting this info. GetPath is giving me the wallet _directory_, excluding the filename, it seems.

Can someone clue me in to the exact semantics / interpretation of the name and path in WalletLocation? Can I just slap "wallet.dat" on the end of the output of GetPath and be good, or are there other cases?

ryanofsky commented:
I'd suggest adding a new method to the [`BerkeleyDatabase`](https://github.com/bitcoin/bitcoin/blob/3515612e069e3730f173a08e60b99f96b174188d/src/wallet/db.h#L107-L196) class:

```diff
+    fs::path DataFilePath() const { return env->Directory() / strFile; }
```

and printing the path with `GetDBHandle().DataFilePath().string()`

gwillen commented:
@ryanofsky I ended up taking a slightly different approach, because at this point we do not have a CWallet yet to get a db handle from (we are in the flow to open one.) Instead, I added a free function for normalizing paths.

(I could have accomplished the same thing without adding a function, by calling SplitWalletPath (but it's static, and a slightly weird interface for this) or GetWalletEnv (but it's a very weird interface for this, and also it manipulates g_dbenvs, which I don't know the consequences of and seems a little heavy-handed for just figuring out a path.)

ryanofsky commented:
Looks good if you s/wallet database file/data file path/ and s/FullWalletFilePath/WalletDataFilePath/.

I'd suggest:

```c++
/** Given a wallet directory path or legacy file path, return path to main data file in the wallet database. */
fs::path WalletDataFilePath(const fs::path& wallet_path);
```

Wallet databases are really directories, not files. At a low level, wallet databases include not just `wallet.dat` files but also write-ahead `database/log.??????????` files. At the user level, the `-wallet` option and `createwallet` rpc create new wallets as directories, and the `-rpcwallet` option and `getwalletinfo` rpc use directory paths.

The only reason it makes sense to reference files instead of directories in log messages is to disambiguate in the legacy case where there can be multiple data files for different wallets stored in the same directory and sharing mixed log files. (This was a misfeature which is mostly removed, except we still support loading these wallets in the top level walletdir.)

gwillen commented:
Got it, thanks @ryanofsky , makes sense. Changed as suggested!

ryanofsky reviewed (approved):
utACK 17d0114bc2e307cdf7feba7d12adc71dfc4f9d6d

It would be good to update the PR description https://github.com/bitcoin/bitcoin/pull/15491#issue-256517395 which still references "[default wallet]".

promag reviewed (commented):
utACK 17d0114.

gwillen commented on src/wallet/db.cpp:
```diff
@@ -84,6 +84,13 @@ bool IsWalletLoaded(const fs::path& wallet_path)
     return database && database->IsDatabaseLoaded(database_filename);
 }
 
+fs::path WalletDataFilePath(const fs::path& wallet_path) {
```
Curses, done.

I'm afraid my brain is stuck in the wrong style for life, sorry.

gwillen commented:
Fixed the PR description.

Fixed the brace @promag was complaining about.

No other changes.

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/15491/commits/faf369880822188f2738c6f046d9ef14c9585713
